databuilder/query_engines/mssql.py

```diff
@@ -61,3 +65,33 @@ def execute_query(self):
             # the normal manner
             with super().execute_query() as results:
                 yield results
+
+    def round_to_first_of_month(self, date):
+        date = type_coerce(date, sqlalchemy_types.Date())
+
+        first_of_month = sqlalchemy.func.dateadd(
+            sqlalchemy.literal_column("DAY"),
+            1,
+            sqlalchemy.func.eomonth(date, -1),
+        )
+
+        return type_coerce(first_of_month, sqlalchemy_types.Date())
+
+    def round_to_first_of_year(self, date):
+        date = type_coerce(date, sqlalchemy_types.Date())
+
+        # set day to first of the month
+        first_of_month = sqlalchemy.func.dateadd(
+            sqlalchemy.literal_column("DAY"),
+            1,
+            sqlalchemy.func.eomonth(date, -1),
+        )
+
+        # set month to first month of the year
+        first_of_year = sqlalchemy.func.dateadd(
+            sqlalchemy.literal_column("MONTH"),
+            -(sqlalchemy.func.month(date) - 1),
+            first_of_month,
+        )
+
+        return type_coerce(first_of_year, sqlalchemy_types.Date())
```

Comment:
We do something simpler [in v1](
https://github.com/opensafely-core/cohort-extractor/blob/main/cohortextractor/date_expressions.py#L325-L326
): 

```
    def date_function_first_day_of_year(self, date):
        return f"DATEFROMPARTS(YEAR({date}), 1, 1)"
```

In fact, seeing this I wonder whether we couldn't do the same for `round_to_first_of_month` too.

@evansd might know why we used `f"DATEADD(DAY, 1, EOMONTH({date}, -1))"` in v1.

Reply:
I have no idea. My only thought is that in the `DATEFROMPARTS` solution the `{date}` expression needs to occur twice and it may itself be a complex expression with a bunch of different nested CASE statements etc. But presumably the database is quite capable of optimising this away, so other than making the SQL hard to read it probably doesn't matter.

Reply:
Rewrote both the MSSQL methods to use `DATEFROMPARTS`, seems a lot cleaner in the code 👍